tox_pdm/plugin4.py

```diff
@@ -0,0 +1,232 @@
+"""Plugin specification for Tox 4"""
+from __future__ import annotations
+
+import os
+import shutil
+import sys
+import typing as t
+from pathlib import Path
+
+from tox.config.cli.parser import DEFAULT_VERBOSITY
+from tox.config.sets import EnvConfigSet
+from tox.execute.api import Execute
+from tox.execute.local_sub_process import LocalSubProcessExecuteInstance
+from tox.execute.request import StdinSource
+from tox.plugin import impl
+from tox.tox_env.package import Package, PackageToxEnv
+from tox.tox_env.python.api import Python, PythonInfo
+from tox.tox_env.python.package import WheelPackage
+from tox.tox_env.python.pip.pip_install import Pip
+from tox.tox_env.python.runner import PythonRun
+from virtualenv.discovery.builtin import get_interpreter
+
+from tox_pdm.utils import clone_pdm_files, get_env_lib_path, is_same_path, is_sub_array
+
+if t.TYPE_CHECKING:
+
+    from argparse import ArgumentParser
+
+    from tox.execute.api import ExecuteInstance, ExecuteOptions
+    from tox.execute.request import ExecuteRequest
+    from tox.execute.stream import SyncWrite
+    from tox.tox_env.api import ToxEnvCreateArgs
+    from tox.tox_env.register import ToxEnvRegister
+
+
+@impl
+def tox_add_option(parser: ArgumentParser) -> None:
+    os.environ["TOX_TESTENV_PASSENV"] = "PYTHONPATH"
+    parser.add_argument("--pdm", default="pdm", help="The executable path of PDM")
+
+
+@impl
+def tox_register_tox_env(register: ToxEnvRegister) -> t.Optional[bool]:
+    register.add_run_env(PdmRunner)
+    register.add_package_env(PdmPackageEnv)
+    register.default_run_env = "pdm"
+
+
+class Pdm(Python):
+    def __init__(self, create_args: ToxEnvCreateArgs) -> None:
+        self._executor: t.Optional[Execute] = None
+        self._installer: t.Optional[Pip] = None
+        self._package_path: t.Optional[Path] = None
+        super().__init__(create_args)
+
+    @property
+    def executor(self) -> Execute:
+        if not self._executor:
+            self._executor = PDMRunExecutor(self.options.is_colored)
+        return self._executor
+
+    @property
+    def installer(self) -> Pip:
+        if self._installer is None:
+            self._installer = PipPep582(self)
+        return self._installer
+
+    @property
+    def package_path(self) -> Path:
+        if not self._package_path:
+            self._package_path = Path(
```

Comment:
you likely can save here a line if you assign the line below to a local variable first